src/components/TodoItem.tsx

```diff
@@ -0,0 +1,127 @@
+import classNames from 'classnames';
+import React, { useEffect } from 'react';
+import { Todo } from '../types/Todo';
+
+type Props = {
+  todo: Todo;
+  loading?: boolean;
+  onDelete?: () => void;
+  onToggle?: () => void;
+  onRename?: (newTitle: string) => Promise<void>;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  loading = false,
+  onDelete = () => {},
+  onToggle = () => {},
+  onRename = () => Promise.resolve(),
+}) => {
+  const [editing, setEditing] = React.useState(false);
+  const [title, setTitle] = React.useState(todo.title);
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (editing) {
+      inputRef.current?.focus();
+    }
+  }, [editing]);
+
+  const handleSubmit = async (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const normalizedTitle = title.trim();
+
+    if (normalizedTitle === todo.title) {
+      setEditing(false);
+
+      return;
+    }
+
+    if (!normalizedTitle) {
+      onDelete();
+
+      return;
+    }
+
+    try {
+      await onRename(normalizedTitle);
+
+      setEditing(false);
+    } catch (error) {
+      inputRef.current?.focus();
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={classNames('todo', { completed: todo.completed })}
+    >
+       {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
```

Comment:
Don't disable linter

Reply:
This is a hint from metor, since otherwise I can't commit